modules/src/core/ics05_port/context.rs

```diff
@@ -1,70 +1,114 @@
-        self.authenticate_capability(&Self::port_capability_name(port_id), capability)
-            .is_ok()
+    /// Get the `Capability` associated with the specified `PortId`
+    fn get_port_capability(&self, port_id: PortId) -> Result<Self::Capability, Error> {
+        self.get_capability(&port_capability_name(port_id))
     }
 
-    fn port_capability_name(port_id: PortId) -> CapabilityName {
-        PortsPath(port_id)
-            .to_string()
-            .parse()
-            .expect("PortsPath cannot be empty string")
+    /// Authenticate a `Capability` against the specified `PortId` by checking if the capability
+    /// was previously generated and bound to the specified port
+    fn authenticate_port_capability(
+        &self,
+        port_id: PortId,
+        capability: &Self::Capability,
+    ) -> Result<(), Error> {
+        self.authenticate_capability(&port_capability_name(port_id), capability)
     }
 }
 
-pub trait PortKeeper: CapabilityKeeper + PortReader {
+pub trait PortCapabilityKeeper<M: Into<ModuleId>>:
+    PortCapabilityReader<M> + CapabilityKeeper<M>
+{
     /// Binds to a port and returns the associated capability
-    fn bind_port(&mut self, port_id: PortId) -> Result<PortCapability, Error> {
+    fn bind_port(
+        &mut self,
+        port_id: PortId,
+    ) -> Result<<Self as CapabilityKeeper<M>>::Capability, Error> {
         if self.is_bound(port_id.clone()) {
             Err(Error::port_already_bound(port_id))
         } else {
-            self.new_capability(Self::port_capability_name(port_id))
+            self.new_capability(port_capability_name(port_id))
                 .map(Into::into)
         }
     }
 }
 
-pub trait CapabilityKeeper {
+pub trait CapabilityKeeper<M: Into<ModuleId>> {
+    /// The concrete associated `Capability` type
+    type Capability: Capability;
+
+    /// The `ModuleId` that this `CapabilityKeeper` is scoped to
+    fn module_id(&self) -> M;
+
     /// Create a new capability with the given name.
     /// Return an error if the capability was already taken.
-    fn new_capability(&mut self, name: CapabilityName) -> Result<Capability, Error>;
+    fn new_capability(&mut self, name: CapabilityName) -> Result<Self::Capability, Error>;
 
     /// Claim the specified capability using the specified name.
     /// Return an error if the capability was already taken.
-    fn claim_capability(&mut self, name: CapabilityName, capability: Capability);
+    fn claim_capability(
+        &mut self,
+        name: CapabilityName,
+        capability: Self::Capability,
+    ) -> Result<(), Error>;
 
     /// Release a previously claimed or created capability
-    fn release_capability(&mut self, name: CapabilityName, capability: Capability);
+    fn release_capability(
+        &mut self,
+        name: CapabilityName,
+        capability: Self::Capability,
+    ) -> Result<(), Error>;
 }
 
-pub trait CapabilityReader {
+pub trait CapabilityReader<M: Into<ModuleId>> {
```

Comment:
This is the main change in this PR. `CapabilityReader` and `CapabilityKeeper` are now generic over `ModuleId`. This follows from the intuition that they are always scoped to a module and the library never makes calls to the underlying host system's (unscoped) `CapabilityReader`. 
I experimented with two other designs - one with a composition style approach (see https://github.com/informalsystems/ibc-rs/pull/2018/commits/ec20b5afcb967e181bb3da7d7e65c34bc07774f2) and another with a freestanding `CapabilityReader` with a blanket scoped implementation, but both those designs proved to be unwieldy. The benefit of this approach is that it makes the least assumptions on the host's capability system and allows users to write a blanket implementation for all modules.